gen/struct_test.go

```diff
@@ -0,0 +1,344 @@
+			}}),
+		},
+		{
+			&testdata.PrimitiveOptionalStruct{BoolField: boolp(true)},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{ID: 1, Value: wire.NewValueBool(true)},
+			}}),
+		},
+		{
+			&testdata.PrimitiveOptionalStruct{ByteField: bytep(1)},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{ID: 2, Value: wire.NewValueI8(1)},
+			}}),
+		},
+		{
+			&testdata.PrimitiveOptionalStruct{Int16Field: int16p(2)},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{ID: 3, Value: wire.NewValueI16(2)},
+			}}),
+		},
+		{
+			&testdata.PrimitiveOptionalStruct{Int32Field: int32p(3)},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{ID: 4, Value: wire.NewValueI32(3)},
+			}}),
+		},
+		{
+			&testdata.PrimitiveOptionalStruct{Int64Field: int64p(4)},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{ID: 5, Value: wire.NewValueI64(4)},
+			}}),
+		},
+		{
+			&testdata.PrimitiveOptionalStruct{DoubleField: doublep(5.0)},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{ID: 6, Value: wire.NewValueDouble(5.0)},
+			}}),
+		},
+		{
+			&testdata.PrimitiveOptionalStruct{StringField: stringp("foo")},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{ID: 7, Value: wire.NewValueString("foo")},
+			}}),
+		},
+		{
+			&testdata.PrimitiveOptionalStruct{BinaryField: []byte("bar")},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{ID: 8, Value: wire.NewValueBinary([]byte("bar"))},
+			}}),
+		},
+	}
+
+	for _, tt := range tests {
+		assert.True(
+			t,
+			wire.ValuesAreEqual(tt.v, tt.s.ToWire()),
+			"%v.ToWire() != %v", tt.s, tt.v,
+		)
+
+		var s testdata.PrimitiveOptionalStruct
+		if assert.NoError(t, s.FromWire(tt.v)) {
+			assert.Equal(t, tt.s, &s)
+		}
+	}
+}
+
+func TestContainersRequiredStruct(t *testing.T) {
+	tests := []struct {
+		s *testdata.ContainersRequiredStruct
+		v wire.Value
+	}{
+		{
+			&testdata.ContainersRequiredStruct{
+				ListOfStrings:      []string{"foo", "bar", "baz"},
+				SetOfInts:          map[int32]struct{}{1: struct{}{}, 2: struct{}{}},
+				MapOfIntsToDoubles: map[int64]float64{1: 2.0, 3: 4.0},
+			},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{
+					ID: 1,
+					Value: wire.NewValueList(wire.List{
+						ValueType: wire.TBinary,
+						Size:      3,
+						Items: wire.ValueListFromSlice([]wire.Value{
+							wire.NewValueString("foo"),
+							wire.NewValueString("bar"),
+							wire.NewValueString("baz"),
+						}),
+					}),
+				},
+				{
+					ID: 2,
+					Value: wire.NewValueSet(wire.Set{
+						ValueType: wire.TI32,
+						Size:      2,
+						Items: wire.ValueListFromSlice([]wire.Value{
+							wire.NewValueI32(1),
+							wire.NewValueI32(2),
+						}),
+					}),
+				},
+				{
+					ID: 3,
+					Value: wire.NewValueMap(wire.Map{
+						KeyType:   wire.TI64,
+						ValueType: wire.TDouble,
+						Size:      2,
+						Items: wire.MapItemListFromSlice([]wire.MapItem{
```

Comment:
does serialization of the map lead to a consistent ordering of map items? Or does it use range to iterate over the map, which could lead to different orders in the resulting wire value.